### PR TITLE
feat: support importing zstd compressed snapshot file directly

### PR DESCRIPTION
**Pull request by hanabi1224**
## Summary of changes

As part of https://github.com/ChainSafe/forest/issues/2741

Changes introduced in this pull request:

- Automatically decompress a snapshot file if it is zstd compressed
- Unittests
- CI sync check tests

## Reference issue to close (if applicable)

Closes

## Change checklist

- [x] I have performed a self-review of my own code,
- [x] I have made corresponding changes to the documentation,
- [x] I have added tests that prove my fix is effective or that my feature works (if possible),
- [x] I have made sure the [CHANGELOG][1] is up-to-date. All user-facing changes should be reflected in this document.

[1]: https://github.com/ChainSafe/forest/blob/main/CHANGELOG.md

### CODE REVIEW

File: scripts/gen_coverage_report.sh [hunk 1]
```diff
@@ -26,10 +26,13 @@ function cov {
...
 
+aria2c -x5 -d "$TMP_DIR" -o calibnet_snapshot.zst https://snapshots.calibrationnet.filops.net/minimal/latest.zst
 cargo llvm-cov --workspace clean
 cargo llvm-cov --workspace --no-report --features slow_tests
 cov forest-cli --chain calibnet db clean --force
 cov forest-cli --chain calibnet snapshot fetch --aria2 -s "$TMP_DIR"
+SNAPSHOT_PATH=$(find "$TMP_DIR" -name \*.zst | head -n 1)
+cov forest --chain calibnet --encrypt-keystore false --import-snapshot "$SNAPSHOT_PATH" --height=-200 --detach --track-peak-rss
 SNAPSHOT_PATH=$(find "$TMP_DIR" -name \*.car | head -n 1)
 cov forest --chain calibnet --encrypt-keystore false --import-snapshot "$SNAPSHOT_PATH" --height=-200 --detach --track-peak-rss
```
Comment (lemmih) [inline scripts/gen_coverage_report.sh:37]: This code tries to detach two instances of Forest. I don't think that will work (and the errors will be hidden in the output files). You probably just want to verify that Forest can load the compressed snapshot.

Comment (lemmih) [reply scripts/gen_coverage_report.sh:37]: You could even clean the DB between the two invocations.

**Review by lemmih — CHANGES_REQUESTED** [review]
Looks good.

File: scripts/gen_coverage_report.sh (hunk 1 above)
Comment (hanabi1224) [reply scripts/gen_coverage_report.sh]: Good catch, fix. Thanks!

**Review by lemmih — APPROVED** [review]

File: utils/forest_utils/src/net/download.rs [hunk 2]
```diff
@@ -107,3 +104,87 @@ impl FetchProgress<BufReader<async_fs::File>> {
...
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &mut [u8],
+    ) -> Poll<std::io::Result<usize>> {
+        if let Some(left) = self.left_mut() {
+            Pin::new(left).poll_read(cx, buf)
+        } else if let Some(right) = self.right_mut() {
+            Pin::new(right).poll_read(cx, buf)
+        } else {
+            unimplemented!("This branch should never be hit")
```
Comment (LesnyRumcajs) [inline utils/forest_utils/src/net/download.rs:187]: I don't think we are planning to implement it :)
```suggestion
            panic!("This branch should never be hit")
```

**Review by LesnyRumcajs — APPROVED** [review]
I like it!

File: utils/forest_utils/src/net/download.rs (hunk 2 above)
Comment (hanabi1224) [reply utils/forest_utils/src/net/download.rs]: Fixed

Comment (hanabi1224) [reply utils/forest_utils/src/net/download.rs]: Hmm, saw this on my phone and it did not show as suggested change : )

**Pull request merged**